# Capstone Project - The Battle of the Neighborhoods (Week 2)
### Applied Data Science Capstone by IBM/Coursera

## Table of contents
* Introduciton : Business Problem
* Data- Part 1 and Part 2
* Methodology
* Analysis
* Result, Discussion
* Conclusion

## Introduction 

Toronto, city, capital of the province of Ontario, southeastern Canada. It is the most populous city in Canada, a multicultural city, and the country’s financial and commercial centre, with a current population of 6,139,000 in 2019.

Similarily it has huge amount of traveller across world.

The aim of this report is to study and analyze the neighborhoods of Toronto city and group them into similar clusters for a travel company who can use the data to inform its traveller regarding popular areas like for shopping, food etc. This will help to save lot of time for traveller as they do not have to search alot for those things and also will help Travel company to gain publicity.

## Lets's get started Getting important libraries loaded

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
%matplotlib inline

## Data - Part 1- wikidata+Location data

Data will be used thorugh html link. This consist Raw data of Bororugh and Neighborhood. The following will be used-

1- Loading wikidata in data frame.<br>
2- Cleaning the data<br>
3- Loading Location data from http://cocl.us/Geospatial_data <br>
4- Joinging both data frame to create one final Data frame I.e. df1_final

In [2]:
#Reading data from Wikipedia using .read method-

canada_df= pd.read_html ("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
candf_2= canada_df[0].dropna()
candf_2


Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [3]:
candf_2.shape

(180, 3)

In [4]:
# Manipulating data to make a good view of Borough and Neighorhood

df1 = candf_2[candf_2['Borough'] != "Not assigned"]
df1.head()
df2 = df1.copy()
df2['Neighborhood'] = df2.apply(
    lambda row: row['Borough'] if row['Neighborhood'] == 'Not assigned' else row['Neighborhood'],
    axis=1
)
df2.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [5]:
#if there is more than one neighborhood in one postal code area, combine those rows and separate the neighborhoods with a comma (group the dataframe by postcode and borough,and then join the neighborhoods)
df3 = df2.copy()
df3 = (df3.groupby(['Postal Code','Borough'])['Neighborhood']
       .apply(lambda x: ','.join(set(x.dropna())))
       .reset_index())

df3 = pd.DataFrame(df3)

df3.tail()

Postal Code    Borough                                       Neighborhood
98          M9N       York                                             Weston
99          M9P  Etobicoke                                          Westmount
100         M9R  Etobicoke  Kingsview Village, St. Phillips, Martin Grove ...
101         M9V  Etobicoke  South Steeles, Silverstone, Humbergate, Jamest...
102         M9W  Etobicoke                Northwest, West Humber - Clairville

In [6]:
# Getting Location Data-

location_df= pd.read_csv('http://cocl.us/Geospatial_data', sep = ',')
location_df.head()


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [7]:
#Using. Concat method to join location and neighborhood data-

df_final= pd.concat([df3, location_df], axis = 1)
df_final.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Postal Code   103 non-null    object 
 1   Borough       103 non-null    object 
 2   Neighborhood  103 non-null    object 
 3   Postal Code   103 non-null    object 
 4   Latitude      103 non-null    float64
 5   Longitude     103 non-null    float64
dtypes: float64(2), object(4)
memory usage: 5.0+ KB


In [8]:
df_final.tail(20)

Postal Code           Borough  \
83          M6R      West Toronto   
84          M6S      West Toronto   
85          M7A  Downtown Toronto   
86          M7R       Mississauga   
87          M7Y      East Toronto   
88          M8V         Etobicoke   
89          M8W         Etobicoke   
90          M8X         Etobicoke   
91          M8Y         Etobicoke   
92          M8Z         Etobicoke   
93          M9A         Etobicoke   
94          M9B         Etobicoke   
95          M9C         Etobicoke   
96          M9L        North York   
97          M9M        North York   
98          M9N              York   
99          M9P         Etobicoke   
100         M9R         Etobicoke   
101         M9V         Etobicoke   
102         M9W         Etobicoke   

                                          Neighborhood Postal Code   Latitude  \
83                              Parkdale, Roncesvalles         M6R  43.648960   
84                                  Runnymede, Swansea         M6S  43.651571   
85         Queen's Park, Ontario Provincial Government         M7A  43.662301   
86               Canada Post Gateway Processing Centre         M7R  43.636966   
87   Business reply mail Processing Centre, South C...         M7Y  43.662744   
88        New Toronto, Mimico South, Humber Bay Shores         M8V  43.605647   
89                              Alderwood, Long Branch         M8W  43.602414   
90       The Kingsway, Montgomery Road, Old Mill North         M8X  43.653654   
91   Old Mill South, King's Mill Park, Sunnylea, Hu...         M8Y  43.636258   
92   Mimico NW, The Queensway West, South of Bloor,...         M8Z  43.628841   
93             Islington Avenue, Humber Valley Village         M9A  43.667856   
94   West Deane Park, Princess Gardens, Martin Grov...         M9B  43.650943   
95   Eringate, Bloordale Gardens, Old Burnhamthorpe...         M9C  43.643515   
96                                       Humber Summit         M9L  43.756303   
97                                    Humberlea, Emery         M9M  43.724766   
98                                              Weston         M9N  43.706876   
99                                           Westmount         M9P  43.696319   
100  Kingsview Village, St. Phillips, Martin Grove ...         M9R  43.688905   
101  South Steeles, Silverstone, Humbergate, Jamest...         M9V  43.739416   
102                Northwest, West Humber - Clairville         M9W  43.706748   

     Longitude  
83  -79.456325  
84  -79.484450  
85  -79.389494  
86  -79.615819  
87  -79.321558  
88  -79.501321  
89  -79.543484  
90  -79.506944  
91  -79.498509  
92  -79.520999  
93  -79.532242  
94  -79.554724  
95  -79.577201  
96  -79.565963  
97  -79.532242  
98  -79.518188  
99  -79.532242  
100 -79.554724  
101 -79.588437  
102 -79.594054

In [9]:
#considering only boroughs that contain the word Toronto
df_toronto = df_final[df_final['Borough'].str.contains('Toronto')].reset_index(drop=True)
df_toronto.head()

Postal Code          Borough                    Neighborhood Postal Code  \
0         M4E     East Toronto                     The Beaches         M4E   
1         M4K     East Toronto    The Danforth West, Riverdale         M4K   
2         M4L     East Toronto  India Bazaar, The Beaches West         M4L   
3         M4M     East Toronto                 Studio District         M4M   
4         M4N  Central Toronto                   Lawrence Park         M4N   

    Latitude  Longitude  
0  43.676357 -79.293031  
1  43.679557 -79.352188  
2  43.668999 -79.315572  
3  43.659526 -79.340923  
4  43.728020 -79.388790

In [10]:
df_toronto.shape

(39, 6)

## Data- Part 2 Contd.

Using Foursquare API to use venue details.

In [11]:
CLIENT_ID = '5SXN5GB3QW2YKIDWFN50SMEIY2AGSLZ5A1SG1SILONJY42BZ' # Foursquare ID
CLIENT_SECRET = 'QCTOHNJVAAA1I0GWHCJ3VTS3NG4UELT0OY0GEW5YOAFGMZ5Y' # Foursquare Secret
VERSION = '20200718' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5SXN5GB3QW2YKIDWFN50SMEIY2AGSLZ5A1SG1SILONJY42BZ
CLIENT_SECRET:QCTOHNJVAAA1I0GWHCJ3VTS3NG4UELT0OY0GEW5YOAFGMZ5Y


In [12]:
import requests
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius


In [13]:
# create a function to get all venues for each neighborhood
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return( nearby_venues)


import requests



In [14]:
#get venues for all toronto neighborhoods and create a dataframe of them
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

In [18]:
toronto_venues.shape


(1634, 7)

In [19]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

## Methodology & Analysis

### Let's begin Analysis part

The following steps are used-
1- After getting Venue details from foursquare API using methods to see each venue details and count.
2- Each neighnorhood will analyzed on the basis of nearby venues and all available places
3- Top venues will be selected to see the popular areas.
4- and finally k-means cluster techinque will be used to find out cluster based on which a final conclusion will be made.


In [20]:
#group the venues across all neighborhoods of toronto
toronto_venues.groupby('Neighborhood').count().head()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     66   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      66   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            57              57   
Brockton, Parkdale Village, Exhibition Place           22              22   
Business reply mail Processing Centre, South Ce...     16              16   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     66              66   

                                                    Venue Longitude  \
Neighborhood                                                          
Berczy Park                                                      57   
Brockton, Parkdale Village, Exhibition Place                     22   
Business reply mail Processing Centre, South Ce...               16   
CN Tower, King and Spadina, Railway Lands, Harb...               16   
Central Bay Street                                               66   

                                                    Venue Category  
Neighborhood                                                        
Berczy Park                                                     57  
Brockton, Parkdale Village, Exhibition Place                    22  
Business reply mail Processing Centre, South Ce...              16  
CN Tower, King and Spadina, Railway Lands, Harb...              16  
Central Bay Street                                              66

In [21]:
nu=len(toronto_venues['Venue Category'].unique()) #number of uniques categories
print('There are {} unique categories.'.format(nu))

There are 234 unique categories.


In [22]:
#create a file with dummy values by venue category
toronto_dummy = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

In [23]:
# add neighborhood column back to dataframe
toronto_dummy['Neighborhood'] = toronto_venues['Neighborhood'] 
# make neighborhood the first column 
fixed_columns = [toronto_dummy.columns[-1]] + list(toronto_dummy.columns[:-1])
toronto_dummy = toronto_dummy[fixed_columns]
toronto_dummy.head()

Yoga Studio  Airport  Airport Food Court  Airport Lounge  Airport Service  \
0            0        0                   0               0                0   
1            0        0                   0               0                0   
2            0        0                   0               0                0   
3            0        0                   0               0                0   
4            0        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   ...  Theater  Theme Restaurant  Toy / Game Store  Trail  Train Station  \
0  ...        0                 0                 0      1              0   
1  ...        0                 0                 0      0              0   
2  ...        0                 0                 0      0              0   
3  ...        0                 0                 0      0              0   
4  ...        0                 0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Women's Store  
0         0              0  
1         0              0  
2         0              0  
3         0              0  
4         0              0  

[5 rows x 234 columns]

In [24]:
#group neighborhoods by taking mean number of each category
torontogroup = toronto_dummy.groupby('Neighborhood').mean().reset_index()
torontogroup.head()

Neighborhood  Yoga Studio  Airport  \
0                                        Berczy Park     0.000000   0.0000   
1       Brockton, Parkdale Village, Exhibition Place     0.000000   0.0000   
2  Business reply mail Processing Centre, South C...     0.000000   0.0000   
3  CN Tower, King and Spadina, Railway Lands, Har...     0.000000   0.0625   
4                                 Central Bay Street     0.015152   0.0000   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0              0.0000           0.000            0.000             0.000   
1              0.0000           0.000            0.000             0.000   
2              0.0000           0.000            0.000             0.000   
3              0.0625           0.125            0.125             0.125   
4              0.0000           0.000            0.000             0.000   

   American Restaurant  Antique Shop  Aquarium  ...  Theater  \
0                  0.0           0.0       0.0  ...      0.0   
1                  0.0           0.0       0.0  ...      0.0   
2                  0.0           0.0       0.0  ...      0.0   
3                  0.0           0.0       0.0  ...      0.0   
4                  0.0           0.0       0.0  ...      0.0   

   Theme Restaurant  Toy / Game Store  Trail  Train Station  \
0               0.0               0.0    0.0            0.0   
1               0.0               0.0    0.0            0.0   
2               0.0               0.0    0.0            0.0   
3               0.0               0.0    0.0            0.0   
4               0.0               0.0    0.0            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                       0.017544               0.0                    0.0   
1                       0.000000               0.0                    0.0   
2                       0.000000               0.0                    0.0   
3                       0.000000               0.0                    0.0   
4                       0.015152               0.0                    0.0   

   Wine Bar  Women's Store  
0  0.000000            0.0  
1  0.000000            0.0  
2  0.000000            0.0  
3  0.000000            0.0  
4  0.015152            0.0  

[5 rows x 234 columns]

In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [27]:
num_top_venues = 5

import numpy as np

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = torontogroup['Neighborhood']



for ind in np.arange(torontogroup.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(torontogroup.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...    Light Rail Station   
3   CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville          Dessert Shop   
9                                    Davisville North        Sandwich Place   
10                       Dufferin, Dovercourt Village                Bakery   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park                  Park   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
15                      High Park, The Junction South    Mexican Restaurant   
16                     India Bazaar, The Beaches West                  Park   
17          Kensington Market, Chinatown, Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                           Little Portugal, Trinity                   Bar   
20                        Moore Park, Summerhill East            Playground   
21                  North Toronto West, Lawrence Park        Clothing Store   
22                             Parkdale, Roncesvalles        Breakfast Spot   
23        Queen's Park, Ontario Provincial Government           Coffee Shop   
24                          Regent Park, Harbourfront           Coffee Shop   
25                           Richmond, Adelaide, King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn                Garden   
28                                 Runnymede, Swansea                  Café   
29                                     St. James Town           Coffee Shop   
30                        St. James Town, Cabbagetown           Coffee Shop   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District                  Café   
33  Summerhill West, Rathnelly, South Hill, Forest...                   Pub   
34                The Annex, North Midtown, Yorkville                  Café   
35                                        The Beaches                 Trail   
36                       The Danforth West, Riverdale      Greek Restaurant   
37           Toronto Dominion Centre, Design Exchange           Coffee Shop   
38                     University of Toronto, Harbord                  Café   

    2nd Most Common Venue          3rd Most Common Venue  \
0            Cocktail Bar                         Bakery   
1             Coffee Shop                 Breakfast Spot   
2           Auto Workshop                  Burrito Place   
3         Airport Service               Airport Terminal   
4      Italian Restaurant                 Sandwich Place   
5                    Café                           Park   
6     Japanese Restaurant               Sushi Restaurant   
7              Restaurant                           Café   
8             Pizza Place                 Sandwich Place   
9                    Park               Department Store   
10               Pharmacy                           Café   
11                   Café                          Hotel   
12                  Trail                  Jewelry Store   
13            Coffee Shop            Jap

In [28]:
neighborhoods_venues_sorted.head(20 )

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...    Light Rail Station   
3   CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville          Dessert Shop   
9                                    Davisville North        Sandwich Place   
10                       Dufferin, Dovercourt Village                Bakery   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park                  Park   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
15                      High Park, The Junction South    Mexican Restaurant   
16                     India Bazaar, The Beaches West                  Park   
17          Kensington Market, Chinatown, Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                           Little Portugal, Trinity                   Bar   

    2nd Most Common Venue          3rd Most Common Venue  \
0            Cocktail Bar                         Bakery   
1             Coffee Shop                 Breakfast Spot   
2           Auto Workshop                  Burrito Place   
3         Airport Service               Airport Terminal   
4      Italian Restaurant                 Sandwich Place   
5                    Café                           Park   
6     Japanese Restaurant               Sushi Restaurant   
7              Restaurant                           Café   
8             Pizza Place                 Sandwich Place   
9                    Park               Department Store   
10               Pharmacy                           Café   
11                   Café                          Hotel   
12                  Trail                  Jewelry Store   
13            Coffee Shop            Japanese Restaurant   
14               Aquarium                          Hotel   
15                   Café                Thai Restaurant   
16         Sandwich Place                    Pizza Place   
17            Coffee Shop          Vietnamese Restaurant   
18               Bus Line                    Swim School   
19  Vietnamese Restaurant  Vegetarian / Vegan Restaurant   

   4th Most Common Venue      5th Most Common Venue  
0               Pharmacy                Cheese Shop  
1                    Gym                    Stadium  
2                    Spa             Farmers Market  
3    Rental Car Location            Harbor / Marina  
4                   Café        Japanese Restaurant  
5            Coffee Shop         Italian Restaurant  
6             Restaurant                    Gay Bar  
7                  Hotel                        Gym  
8                    Gym           Sushi Restaurant  
9   Gym / Fitness Center                Pizza Place  
10      Recording Studio  Middle Eastern Restaurant  
11                   Gym        Japanese Restaurant  
12      Sushi Restaurant           Department Store  
13    Italian Restaurant             Cosmetics Shop  
14                  Café        Sporting Goods Shop  
15             Bookstore        Fried Chicken Joint  
16      Sushi Restaurant                    Brewery  
17                   Bar         Mexican Restaurant  
18        Farmers Market                Event Space  
19           Coffee Shop                

In [29]:
# set number of clusters
kclusters = 7

toronto_grouped_clustering = torontogroup.drop('Neighborhood', 1)

from sklearn.cluster import KMeans

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:7]

array([0, 0, 0, 0, 0, 0, 0])

In [38]:
toronto_merged = df_toronto

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(40)

Postal Code           Borough  \
0          M4E      East Toronto   
1          M4K      East Toronto   
2          M4L      East Toronto   
3          M4M      East Toronto   
4          M4N   Central Toronto   
5          M4P   Central Toronto   
6          M4R   Central Toronto   
7          M4S   Central Toronto   
8          M4T   Central Toronto   
9          M4V   Central Toronto   
10         M4W  Downtown Toronto   
11         M4X  Downtown Toronto   
12         M4Y  Downtown Toronto   
13         M5A  Downtown Toronto   
14         M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
16         M5E  Downtown Toronto   
17         M5G  Downtown Toronto   
18         M5H  Downtown Toronto   
19         M5J  Downtown Toronto   
20         M5K  Downtown Toronto   
21         M5L  Downtown Toronto   
22         M5N   Central Toronto   
23         M5P   Central Toronto   
24         M5R   Central Toronto   
25         M5S  Downtown Toronto   
26         M5T  Downtown Toronto   
27         M5V  Downtown Toronto   
28         M5W  Downtown Toronto   
29         M5X  Downtown Toronto   
30         M6G  Downtown Toronto   
31         M6H      West Toronto   
32         M6J      West Toronto   
33         M6K      West Toronto   
34         M6P      West Toronto   
35         M6R      West Toronto   
36         M6S      West Toronto   
37         M7A  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood Postal Code   Latitude  \
0                                         The Beaches         M4E  43.676357   
1                        The Danforth West, Riverdale         M4K  43.679557   
2                      India Bazaar, The Beaches West         M4L  43.668999   
3                                     Studio District         M4M  43.659526   
4                                       Lawrence Park         M4N  43.728020   
5                                    Davisville North         M4P  43.712751   
6                   North Toronto West, Lawrence Park         M4R  43.715383   
7                                          Davisville         M4S  43.704324   
8                         Moore Park, Summerhill East         M4T  43.689574   
9   Summerhill West, Rathnelly, South Hill, Forest...         M4V  43.686412   
10                                           Rosedale         M4W  43.679563   
11                        St. James Town, Cabbagetown         M4X  43.667967   
12                               Church and Wellesley         M4Y  43.665860   
13                          Regent Park, Harbourfront         M5A  43.654260   
14                           Garden District, Ryerson         M5B  43.657162   
15                                     St. James Town         M5C  43.651494   
16                                        Berczy Park         M5E  43.644771   
17                                 Central Bay Street         M5G  43.657952   
18                           Richmond, Adelaide, King         M5H  43.650571   
19  Harbourfront East, Union Station, Toronto Islands         M5J  43.640816   
20           Toronto Dominion Centre, Design Exchange         M5K  43.647177   
21                     Commerce Court, Victoria Hotel         M5L  43.648198   
22                                           Roselawn         M5N  43.711695   
23    Forest Hill North & West, Forest Hill Road Park         M5P  43.696948   
24                The Annex, North Midtown, Yorkville         M5R  43.672710   
25                     University of Toronto, Harbord         M5S  43.662696   
26          Kensington Market, Chinatown, Grange Park         M5T  43.653206   
27  CN Tower, King and Spadina, Railway Lands, Har...         M5V  43.628947   
28                                     Stn A PO Boxes         M5W  43.646435   
29             First Canadian Place, Underground city         M5X  43.648429   
30                                           Christie         M6G  43.669542   
31

In [34]:
# Getting toronto latitude and longitude thorugh google search
latitude=43.653963
longitude= -79.387207

import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=13)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=10,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Results and Discussion

Based on the above analysis, it can be suggested to Traveller company that neighborhood Kensington Market, Chinatown, Grange Park having good resturants 

M5T	Downtown Toronto	Kensington Market, Chinatown, Grange Park	M5T	43.653206	-79.400049	4	Café	Coffee Shop	Vietnamese Restaurant	Bar	Mexican Restaurant

whereas east Toronto is mainly for shopping places-

M5N	Central Toronto	Roselawn	M5N	43.711695	-79.416936	0	Garden	Women's Store	Department Store	Event Space	Ethiopian Restaurant

## Conclusion

Purpose of this report was to anlayze various neighhood and find out the popular places. So that Traveller company can use the data and suggest its customer to directly to explore those place as per their needs.

Final decision shows that Toronto as a city has diverse culture as well as places. It shows Downtown Toronto is mostly can be used for food lovers on one hand East Toronto can be explored for shopping. This gives a clear pictures to travellers who are very new to big city like Toronto can explore the area easily.